lang/funcs/core/os/exec_func.go

```diff
@@ -0,0 +1,164 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package coreos
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os/exec"
+	"sync"
+
+	"github.com/purpleidea/mgmt/lang/funcs"
+	"github.com/purpleidea/mgmt/lang/interfaces"
+	"github.com/purpleidea/mgmt/lang/types"
+)
+
+func init() {
+	funcs.ModuleRegister(ModuleName, "exec", func() interfaces.Func { return &ExecFunc{} })
+}
+
+// ExecFunc runs a string as a shell command, then produces each line from
+// stdout. If the input string changes, then the commands are executed one after
+// the other and the concatenation of their outputs is produced line by line.
+//
+// Note that in the likely case in which the process emits several lines one
+// after the other, the downstream resources might not run for every line unless
+// the "Meta:realize" metaparam is set to true.
+type ExecFunc struct {
+	init *interfaces.Init
+
+	// Used to shutdown the node, killing the current process if needed.
+	// We only run one process at a time.
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// ArgGen returns the Nth arg name for this function.
+func (obj *ExecFunc) ArgGen(index int) (string, error) {
+	seq := []string{"shell_command"}
+	if l := len(seq); index >= l {
+		return "", fmt.Errorf("index %d exceeds arg length of %d", index, l)
+	}
+	return seq[index], nil
+}
+
+// Validate makes sure we've built our struct properly. It is usually unused for
+// normal functions that users can use directly.
+func (obj *ExecFunc) Validate() error {
+	return nil
+}
+
+// Info returns some static info about itself.
+func (obj *ExecFunc) Info() *interfaces.Info {
+	return &interfaces.Info{
+		Pure: false, // definitely false
+		Memo: false,
+		Sig:  types.NewType("func(shell_command str) str"),
+		Err:  obj.Validate(),
+	}
+}
+
+// Init runs some startup code for this function.
+func (obj *ExecFunc) Init(init *interfaces.Init) error {
+	obj.init = init
+	obj.ctx, obj.cancel = context.WithCancel(context.Background())
+	return nil
+}
+
+// Stream returns the changing values that this func has over time.
+func (obj *ExecFunc) Stream() error {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	// Close the output chan to signal that no more values are coming.
+	defer close(obj.init.Output)
+
+	// Kill the current process if it's still running.
+	defer obj.cancel()
+
+	for {
+		select {
+		case input, more := <-obj.init.Input:
+			if !more {
+				return nil
+			}
+			shellCommand := input.Struct()["shell_command"].Str()
+
+			// Run the command, connecting it to ctx so we can kill
+			// it if needed, and to two Readers so we can read its
+			// stdout and stderr.
+			cmd := exec.CommandContext(obj.ctx, "sh", "-c", shellCommand)
```

Comment:
What's wrong with using the shell? Are you concerned about security or performance?

I used the `sh` in order to make sure I am splitting the algorithm the same way the shell does. In particular, the linked code splits on whitespace, which is much simpler than what the shell does; on the input `foo "bar baz" 'hello world'`, the linked code produces `foo`, `"bar`, `baz"`, `'hello`, and `world'`, while the shell would produce `foo`, `bar baz`, and `hello world`.

Using the shell also provides more functionality to the user. For example, if you have a command like `bc` which expects its input from stdin rather than the command line, the with the shell the user can still use `os.exec(fmt.printf("echo %d %% 10 | bc", datetime.now()))` to control `bc`'s stdin.

In most programming languages, there are typically at least two functions for executing outside processes. This derives from C, where there is both `system` and the `execv` family of functions. High level languages often prefer to spawn a separate process than to replace the current process, as a result the second function is often called something other than "exec". But, like in C, languages typically provide one function with a simple API consisting of a single string input, often called "system", but sometimes also "shell" or "command", and a more comprehensive function which takes a list of strings, doesn't delegate to the shell, and allows the user to control other things such as a map of environment variables, a string to use for stdin, and an option to control what to do with stderr and with the exit code.

Since it would make sense to implement the simple function before the complex function, how about keeping the current shell-based implementation and renaming `exec` to "system" or "shell", thus keeping the name "exec" available for a future, more complex version?

Reply:
> Since it would make sense to implement the simple function before the complex function, how about keeping the current shell-based implementation and renaming exec to "system" or "shell", thus keeping the name "exec" available for a future, more complex version?

It's a really good idea. As it happens the exec resource `exec` can easily have flags, and it has the `Shell` flags to let you specify a shell. https://github.com/purpleidea/mgmt/blob/3cea4223657cdbb32fcb0afc54bb736388dfbd5e/engine/resources/exec.go#L64
But assuming you are talking about this function, then yes, we can rename it as you suggest to `system` I think for the shell version.